fix: preserve payment sessions during certain Stripe errors for webhook reconciliation

Saryazdi-Saman opened the pull request:
### Issue
When creating a payment intent with specific parameters (off_session: true, confirm: true, capture_method: "automatic"), the payment module throws errors on certain Stripe responses. This triggers a cleanup function that deletes the payment session.

This creates a significant problem in the payment flow: if the issue resolves and a webhook event is later received from Stripe, the payment session no longer exists to be matched and updated.

Additionally, the payment module currently doesn't update the payment_session status when calling authorizePaymentSession, which creates inconsistency in the payment state.

### Solution
This PR implements several improvements to make payment processing more robust:

1. **Better error handling**: Errors from Stripe are now handled appropriately:
   - If the error is retry-able (network errors, rate limits), a retry is triggered
   - If the issue can't be solved instantly, the error details are stored in payment_session.data
   - This preserves the payment_session for webhook reconciliation or future resolution

2. **Retry mechanism**: Added automatic retry with exponential backoff for errors that might be fixed after retry (connection issues, rate limits)

3. **Status consistency**: Modified authorizePaymentSession method to update payment_session.status regardless of whether the authorization is successful

## Benefits
This change improves payment reliability by:
- Preventing data loss during transient errors
- Maintaining a consistent record for webhook reconciliation 
- Enabling better debugging of payment issues
- Ensuring payment status is always reflected accurately

Note: I haven't included tests yet as I wanted to get feedback on the approach first. I'm happy to add tests once the solution direction is confirmed.

Saryazdi-Saman committed:
fix: preserve payment sessions during certain Stripe errors for webhook reconciliation

fix: add retry mechanism for errors that might be fixed after retry

fix: authorizePaymentSession method will update payment_session.status regardless regardless of wether or not the authorization is successful

sradevski commented on packages/modules/providers/payment-stripe/src/types/index.ts:
```diff
@@ -21,6 +23,10 @@ export interface StripeOptions {
   paymentDescription?: string
 }
 
+export type HandledErrorType = 
```
Since this is just an internal data type, it should live inside the Stripe payment provider (i.e. you can put it at the top of `stripe-base`.

sradevski commented on packages/modules/providers/payment-stripe/src/core/stripe-base.ts:
```diff
@@ -115,6 +116,87 @@ abstract class StripeBase extends AbstractPaymentProvider<StripeOptions> {
   }
 
+  handleStripeError(error: any): HandledErrorType | null {
+    if (error instanceof Stripe.errors.StripeError) {
```
I think we can remove the `if` check as the `switch` will serve a similar purpose (it already checks for stripe-explicit errors, and the default is to throw a generic error)

Saryazdi-Saman replied:
I removed the if check, just switching over types and falling back on default

sradevski commented on packages/modules/providers/payment-stripe/src/core/stripe-base.ts:
```diff
@@ -115,6 +116,87 @@ abstract class StripeBase extends AbstractPaymentProvider<StripeOptions> {
+        currentAttempt <= maxRetries
+      ) {
+        const delay = baseDelay * Math.pow(2, currentAttempt - 1) * (0.5 + Math.random() * 0.5);
+        await new Promise(resolve => setTimeout(resolve, delay));
```
You can do `import { setTimeout } from "timers/promises"`

and then just do `await setTimeout(delay)` here.

Saryazdi-Saman replied:
Great! Done!

sradevski commented on packages/modules/providers/payment-stripe/src/core/stripe-base.ts:
```diff
@@ -115,6 +116,87 @@ abstract class StripeBase extends AbstractPaymentProvider<StripeOptions> {
+        case 'StripeRateLimitError':
+          return {
+            retry: true,
+            data: null,
```
We can just return `{ retry: true }` thereby data being `undefined`.

Saryazdi-Saman replied:
I modified the type definition to simplify it as you suggested. this line is now { retry: true}

sradevski commented on packages/modules/providers/payment-stripe/src/core/stripe-base.ts:
```diff
@@ -115,6 +116,87 @@ abstract class StripeBase extends AbstractPaymentProvider<StripeOptions> {
+            data: null,
+          }
+        case 'StripeAPIError': {
+          return null
```
Shouldn't this just throw? We can let it fallback to the default case

Saryazdi-Saman replied:
I should have left a descriptive comment on this line to explain the reasoning.

According to [Stripe API errors documentation](https://docs.stripe.com/error-handling?lang=node#api-errors), StripeAPIError should be treated as indeterminate - we shouldn't assume it succeeded or failed, and should rely on webhooks for the outcome.

Throwing an error here would trigger the cleanup function that deletes the payment session. By returning null instead, we preserve the payment session, which is critical for when webhooks arrive later with the actual outcome (which could be successful).

This approach follows Stripe's best practice of relying on webhooks for the final state in indeterminate cases. The user can still manually throw an error if payment_session.data is null if they prefer different behavior.

What are your thoughts?

sradevski replied:
Got it, thanks for explaining, I think it makes sense. I just added a small suggestion on how we can improve it.

sradevski commented on packages/modules/providers/payment-stripe/src/core/stripe-base.ts:
```diff
@@ -173,22 +255,21 @@ abstract class StripeBase extends AbstractPaymentProvider<StripeOptions> {
-      id: sessionData.id,
-      data: sessionData,
-    }
+      if (sessionData === null) {
```
You can remove this if check and only return the data below
```
{
        id: sessionData.id,
        data: sessionData as unknown as Record<string, unknown>
}
```

Since if sessionData is not defined we should throw as the payment intent wasn't created successfully

Saryazdi-Saman replied:
I have removed this if check, and simplified it in a one liner.
Before i remove the null possibilities, I need to know your thoughts on returning null as a valid value for cases where result is uncertain and requires webhook updates.

sradevski commented on packages/modules/providers/payment-stripe/src/core/stripe-base.ts:
```diff
@@ -115,6 +116,87 @@ abstract class StripeBase extends AbstractPaymentProvider<StripeOptions> {
+    try {
+      return await apiCall();
+    } catch (error) {
+      const handledError = this.handleStripeError(error)
```
With the suggested changes, this should always be defined, with either `{ retry: false, data: somedata}` or `{retry: true}`, so we can remove the if check below

Saryazdi-Saman replied:
Applied.

sradevski commented on packages/modules/providers/payment-stripe/src/core/stripe-base.ts:
```diff
@@ -115,6 +116,87 @@ abstract class StripeBase extends AbstractPaymentProvider<StripeOptions> {
+        handledError.retry && 
+        currentAttempt > maxRetries
+      ) {
+        return null
```
We should throw here

Saryazdi-Saman replied:
The purpose was to avoid throwing if the result is uncertain so the payment_session wont be deleted in case a webhook event arrives that addresses that payment session.
I'll wait for your feedback before applying any changes

sradevski replied:
I think at this point we can be certain the payment intent wasn't created, or it is highly unlikely, so it's safer to throw.

sradevski commented on packages/modules/providers/payment-stripe/src/core/stripe-base.ts:
```diff
@@ -115,6 +116,87 @@ abstract class StripeBase extends AbstractPaymentProvider<StripeOptions> {
+        return null
+      }
+
+      if (handledError.data) {
```
This check should go above
```
 if (
        handledError.retry &&
        currentAttempt <= maxRetries
      ) {
```

If the handleError returned some data, then return that and complete. Otherwise we retry

Saryazdi-Saman replied:
Perfect, this check is moved to the top to return immediately if there is data.

sradevski reviewed (commented):
Thanks a lot for the contribution! I left some comments on how we can improve it, and once they're addressed we can get it merged

Saryazdi-Saman committed:
Refactor: improve handling structure and syntax

-Move HandledErrorType definition to the top of stripe-base

- Use timers/promises for setTimeout

- Removed data in HandledErrorType when retry is true

Saryazdi-Saman committed:
refactor: improve error handling flow and logic

- Simplify return statement in initiatePayment to handle null cases

- Remove redundant if-check in handleStripeError and rely on switch

- Reorder conditional checks in executeWithRetry for clearer flow

- Update executeWithRetry to check for retry=false condition first

Saryazdi-Saman committed:
clean up

sradevski commented on packages/modules/providers/payment-stripe/src/core/stripe-base.ts:
```diff
@@ -115,6 +119,87 @@ abstract class StripeBase extends AbstractPaymentProvider<StripeOptions> {
+        // API errors should be treated as indeterminate per Stripe documentation
+        // We return null to preserve the payment session and rely on webhooks
+        // for the actual outcome rather than assuming failure
+        return null
```
I would still maybe return
```
{
  retry: false,
  data: {
    state: "indeterminate_error"
  }
}
```

or something like that so at least we have a trace of what happened in the DB

Saryazdi-Saman replied:
You're right, returning a state is better than returning null. I've implemented this change to preserve a trace of what happened

sradevski commented on packages/modules/providers/payment-stripe/src/core/stripe-base.ts:
```diff
@@ -115,6 +119,87 @@ abstract class StripeBase extends AbstractPaymentProvider<StripeOptions> {
+      return await apiCall();
+    } catch (error) {
+      const handledError = this.handleStripeError(error)
+      if (handledError === null) {
```
If we do the suggestion above, we can remove this check.

sradevski reviewed (commented):
Final comment, otherwise LGTM

sradevski commented on packages/modules/providers/payment-stripe/src/core/stripe-base.ts:
```diff
@@ -115,6 +119,87 @@ abstract class StripeBase extends AbstractPaymentProvider<StripeOptions> {
+      // Retries are exhausted or result is null
+      // handleStripeError() didn't throw an error
+      // return null to avoid triggering the cleanup function 
+      return null
```
I think if we get up to this point, it means the payment intent wasn't created (i.e. the error wasn't StripeAPIError, nor was it something where we know the payment intent was created). I think we should throw in this case.

Saryazdi-Saman replied:
I agree. I think I was overcorrecting. At this point it is safe to assume payment intent wasn't created.
Changes are applied.
Let me know if there are any other concerns to address.

Saryazdi-Saman committed:
fix: improve payment error handling and traceability

- Return structured error state for StripeAPIError instead of null

- Throw error when retries are exhausted and no payment intent exists

- Update type definitions to support error state tracking

Saryazdi-Saman committed:
fix formatting and naming
